### ISSUE TITLE
Action processing queues

### ISSUE THREAD

**Issue by ai** [logux/logux#54]
Right now Logux Server process all actions “in parallel” (on one thread, just in async “parallel”). It could cause a problem when actions are related:

1. User A created a post
2. User A posted a comment to Post (but step 1 is not finished yet).

**Comment by ai** [comment]
Solution:

1. Add the `queue` option to `Server#type` and `Server#channel`. By default, it will be `main`.
2. For every client and `queue`, we are creating action queue. The server will start `action`, `resend`, `process`, or `load` only if the previous action in the same queue was finally processed.

If the developer wants to speed-up processing, the developer can put isolated actions to separated `queue`.

Since it is advanced optimization, I think for now e will not add the API to back-end protocol.

**Comment by ai** [comment]
This feature will require big action processing refactoring with proper queues.

**Comment by damingo** [comment]
It appears to be that for optimistic UIs this can cause errors whenever a client reconnects to the server and the client made changes while being offline.
When a client creates - and then edits an item while offline, errors are raised when changes are uploaded to the server when the connection is back. 
As for example "items/create" and "items/edit" are executed in parallel, the edit action does not find the created item as "items/create" (and DB transaction) is not yet finished.

A simple `await setTimeout(200, 'resolved')` at the start of the edit action resolves the problem - but is obviously no solution.

Is something like https://github.com/rogierschouten/async-lock the way to go until queues are implemented?

**Comment by ai** [comment]
Yes. Now, you need to create queue on your own in `server.type()`’s `process` callback.

You are welcome to send PR, we have API design above.  It could be done in a few days. I can answer any questions here or in our [Gitter chat](https://gitter.im/logux/logux).

**Comment by damingo** [comment]
Thanks! For the API design you write 
> The server will start action, resend, process, or load only if the previous action in the same queue was finally processed.

Should "action" be "access" or are you referring to something else? The "access" callback should also be included.

Would the use of https://github.com/mcollina/fastq be acceptable / recommended? 

Not yet sure if / when I get to make a PR.

**Comment by ai** [comment]
> Should "action" be "access" or are you referring to something else? The "access" callback should also be included.

Yeap, `access` should be called in queue too. But we have two problems.

**Problem 1: slow synchronization**

We are calling `access` before adding the action to server’s log. It means, that client will not get `synchronized` status until we call all `access`. In queue, it will slow down `syncing` status for a few minutes.

But maybe it is not a problem. With Optimistic UI sync time is not a big problem.

**Problem 2: action duplicated**

Now, server sends back `synced` command (that all actions were saved on server) only when all `access` callbacks was called. With `access` queue, it could take a time.

If the client loses Internet connection during the actions' synchronization, without `synced` response, the client will try to send all actions back to the server.

I am not sure about the best way to solve it.

> Would the use of https://github.com/mcollina/fastq be acceptable / recommended?

Why not

**Comment by damingo** [comment]
**Problem 1: slow synchronization**

Not much of a problem I think. Logux is fast and adds minimal overhead. As `access` callbacks must be executed in sequence for related actions they can't be taken out of the queue. And as you mentioned, developers can move actions into separate queues to speed things up. They can use as many queues as their related actions allow.

For isolated actions the API should probably offer to execute actions in parallel (as it is done right now). Maybe by setting `queue: null`. As proposed, the default should be that actions are added to the clients `main` queue. As a developer one assumes actions to be processed in sequence, and bugs caused by parallel processing can easily go unnoticed.

**Problem 2: action duplicated**

Maybe sending back `synced` in combination with the actions ID whenever the `access` callback of an action finished?

**Comment by ai** [comment]
> Maybe sending back synced in combination with the actions ID whenever the access callback of an action finished?

I am OK with chancing protocol a little.

But we have another problem:

1. Right now, `sync` uses `added` index (action’s per-node sequence number)
2. We are using `added` in `synced` to simplify detection of what actions we need to sync next time. We can’t keep the full list of synced IDs because it will take a lot of memory
3. Because of `queue` option action could be processed in a different order on server. So we can process an action with bigger `added` index before the action with smaller `added` index

Possible solution: we can change a way how we are putting action to the server log.

1. On `sync` message we will create an array of actions in the memory of server and return `synced` to the client
2. Then we will process actions with queues

It may need to extend Logux Core to add support for this behavior.

**Comment by damingo** [comment]
> Possible solution: we can change a way how we are putting action to the server log.
> 1. On sync message we will create an array of actions in the memory of server and return synced to the client
> 2. Then we will process actions with queues

Seems good to me.
